ip-utils.c

```diff
@@ -12,13 +12,45 @@
  */
 
 #include <errno.h>
+#include <sys/socket.h>
 #include <sys/types.h>
 #include <arpa/inet.h>
+#include <netdb.h>
 #include <net/if.h>
 #include <string.h>
+#include <unistd.h>
+
+#include <glib.h>
 
 #include "ip-utils.h"
 
+int janus_network_detect_local_ip(char *buf, size_t buflen) {
```

Comment:
> so I'd still keep a string as the return value somehow to keep things easier.

That is actually what makes the code *harder* to use because now the caller has to pick the correct buffer size to use (which is one opportunity for errors) and should zero it (which is another). You simplify the API by taking away the possibility for making the wrong decisions, so pushing towards the proper structs actually helps. For example this is how you would use the function to obtain the string afterwards:

```
    janus_network_address a;
    janus_network_address_string_buffer b;
    int error = janus_network_detect_local_ip(janus_ip_network_query_options_any_ip, &a);
    if(error) {
       // handle it.
    }
    janus_network_address_to_string_buffer(&a, &b);
    const char * formatted_ip_address = janus_network_address_string_from_buffer(&b);
    if(formatted_ip_address) {
        // use it.
        printf("Your IP address is: %s\n", formatted_ip_address);
    }
```

Now if the argument is that the primary (or sole) use of the function is to obtain such strings, then that calls for a higher level convenience function which does precisely that:
```
    const char *janus_network_detect_local_ip_as_string(janus_ip_network_query_options addr_type);
```

It's body would be equivalent to this:
```
    const char *janus_network_detect_local_ip_as_string(janus_ip_network_query_options addr_type) {
        janus_network_address a;
        janus_network_address_string_buffer b;
        return janus_network_detect_local_ip(addr_type, &a) || janus_network_address_to_string_buffer(&a, &b) ? NULL : janus_network_address_string_from_buffer(&b);
    }
```

That way you keep the API clean and unambiguous, plus you obtain the convenience function which you were really after.